File: sdk/csharp/Udb.Client/UdbAuthClient.cs
using System.Security.Cryptography;
using System.Text;
using Grpc.Core;
using Grpc.Net.Client;
using AuthnV1 = udb.core.Authn.Services.V1;
using AuthzV1 = udb.core.Authz.Services.V1;

namespace Udb.Client;

/// <summary>
/// Hand-written auth ergonomics over the generated AuthnService / AuthzService
/// clients, mirroring <see cref="UdbClient"/>'s metadata convention (item 110).
/// </summary>
public sealed class UdbAuthClient : IAsyncDisposable
{
    private readonly GrpcChannel? _ownedChannel;
    private readonly UdbMetadata _metadata;
    private readonly UdbCredentials _credentials;
    private readonly AuthnV1.AuthnService.AuthnServiceClient _authn;
    private readonly AuthzV1.AuthzService.AuthzServiceClient _authz;
    private readonly AuthzCache _cache;
    private readonly string? _bundleSecret;

    /// <summary>Standalone constructor: owns its channel and a fresh decision cache.</summary>
    /// <param name="bundleSecret">
    /// When set, signed policy bundles returned by <see cref="GetPolicyBundleAsync"/>
    /// are HMAC-SHA256 verified against this shared secret.
    /// </param>
    public UdbAuthClient(string address, UdbMetadata metadata, AuthzCache? cache = null, string? bundleSecret = null, UdbCredentials? credentials = null)
    {
        _ownedChannel = GrpcChannel.ForAddress(address);
        _metadata = metadata;
        _credentials = credentials ?? new UdbCredentials(metadata.BearerToken, metadata.ApiKey);
        _authn = new AuthnV1.AuthnService.AuthnServiceClient(_ownedChannel);
        _authz = new AuthzV1.AuthzService.AuthzServiceClient(_ownedChannel);
        _cache = cache ?? new AuthzCache();
        _bundleSecret = bundleSecret;
    }

    /// <summary>
    /// Facade constructor: reuses a channel owned elsewhere (the channel is NOT
    /// disposed by this instance) and a shared decision cache.
    /// </summary>
    internal UdbAuthClient(GrpcChannel sharedChannel, UdbMetadata metadata, AuthzCache cache, string? bundleSecret = null, UdbCredentials? credentials = null)
    {
        _ownedChannel = null;
        _metadata = metadata;
        _credentials = credentials ?? new UdbCredentials(metadata.BearerToken, metadata.ApiKey);
        _authn = new AuthnV1.AuthnService.AuthnServiceClient(sharedChannel);
        _authz = new AuthzV1.AuthzService.AuthzServiceClient(sharedChannel);
        _cache = cache;
        _bundleSecret = bundleSecret;
    }

    /// <summary>
    /// Test constructor: injects pre-built generated clients (e.g. backed by a
    /// fake <see cref="Grpc.Core.CallInvoker"/>) so request/header construction
    /// can be exercised without a live server.
    /// </summary>
    internal UdbAuthClient(
        AuthnV1.AuthnService.AuthnServiceClient authn,
        AuthzV1.AuthzService.AuthzServiceClient authz,
        UdbMetadata metadata,
        AuthzCache? cache = null,
        string? bundleSecret = null,
        UdbCredentials? credentials = null)
    {
        _ownedChannel = null;
        _metadata = metadata;
        _credentials = credentials ?? new UdbCredentials(metadata.BearerToken, metadata.ApiKey);
        _authn = authn;
        _authz = authz;
        _cache = cache ?? new AuthzCache();
        _bundleSecret = bundleSecret;
    }

    /// <summary>The shared TTL decision cache backing <see cref="CanAsync"/>/<see cref="RequireAsync"/>.</summary>
    public AuthzCache Cache => _cache;

    /// <summary>Raw generated Authn client, for RPCs not yet wrapped.</summary>
    public AuthnV1.AuthnService.AuthnServiceClient Authn => _authn;

    /// <summary>Raw generated Authz client, for RPCs not yet wrapped.</summary>
    public AuthzV1.AuthzService.AuthzServiceClient Authz => _authz;

    private Metadata Headers()
    {
        var headers = new Metadata
        {
            { "x-tenant-id", _metadata.TenantId },
            { "x-user-id", _metadata.UserId },
            { "x-purpose", _metadata.Purpose },
            { "x-correlation-id", _metadata.CorrelationId },
            { "x-scopes", string.Join(",", _metadata.Scopes) },
            { "x-service-identity", _metadata.ServiceIdentity },
            { "x-udb-project-id", _metadata.ProjectId },
            { "x-udb-client-catalog-version", _metadata.ClientCatalogVersion },
        };
        var bearer = _credentials.BearerToken;
        if (!string.IsNullOrWhiteSpace(bearer))
        {
            headers.Add("authorization", $"Bearer {bearer}");
        }
        var apiKey = _credentials.ApiKey;
        if (!string.IsNullOrWhiteSpace(apiKey))
        {
            headers.Add("x-api-key", apiKey);
        }
        return headers;
    }

    /// <summary>The shared credentials holder backing this client's outbound auth headers.</summary>
    public UdbCredentials Credentials => _credentials;

    // ── Authentication ──────────────────────────────────────────────────────
    public Task<AuthnV1.AuthnResponse> AuthenticateAsync(AuthnV1.AuthnRequest request, CancellationToken ct = default)
        => _authn.AuthenticateAsync(request, Headers(), cancellationToken: ct).ResponseAsync;

    public Task<AuthnV1.AuthnResponse> AuthenticateBearerAsync(string token, CancellationToken ct = default)
        => AuthenticateAsync(new AuthnV1.AuthnRequest { BearerToken = token, TenantHint = _metadata.TenantId, ProjectHint = _metadata.ProjectId }, ct);

    public Task<AuthnV1.AuthnResponse> AuthenticateApiKeyAsync(string apiKey, CancellationToken ct = default)
        => AuthenticateAsync(new AuthnV1.AuthnRequest { ApiKey = apiKey, TenantHint = _metadata.TenantId, ProjectHint = _metadata.ProjectId }, ct);

    public Task<AuthnV1.AuthnResponse> AuthenticateSessionAsync(string sessionId, CancellationToken ct = default)
        => AuthenticateAsync(new AuthnV1.AuthnRequest { SessionId = sessionId, TenantHint = _metadata.TenantId, ProjectHint = _metadata.ProjectId }, ct);

    /// <summary>
    /// Exchange a refresh token (and optional session id) for a fresh access
    /// token via the Authn RefreshToken RPC. Returns the new access token and
    /// its lifetime in seconds.
    /// </summary>
    public async Task<AuthnV1.RefreshTokenResponse> RefreshTokenAsync(
        string refreshToken, string sessionId = "", CancellationToken ct = default)
    {
        var request = new AuthnV1.RefreshTokenRequest { RefreshToken = refreshToken };
        if (!string.IsNullOrEmpty(sessionId))
        {
            request.SessionId = sessionId;
        }
        return await _authn.RefreshTokenAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
    }

    // ── Authorization ─────────────────────────────────────────────────────────
    public async Task<AuthzV1.Decision> AuthorizeAsync(AuthzV1.AuthzRequest request, CancellationToken ct = default)
    {
        var resp = await _authz.AuthorizeAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
        return resp.Decision;
    }

    /// <summary>
    /// Check whether the configured principal may perform <paramref name="action"/>
    /// on <paramref name="resource"/>. The decision is served from (and stored in)
    /// the TTL cache. <paramref name="scopes"/> augments the metadata scopes for
    /// both the request's <c>requested_scopes</c> and the cache key.
    /// </summary>
    public async Task<(bool Allowed, AuthzV1.Decision Decision)> CanAsync(
        AuthzV1.ResourceRef resource, string action, string purpose = "",
        IEnumerable<string>? scopes = null, CancellationToken ct = default)
    {
        var effectivePurpose = string.IsNullOrEmpty(purpose) ? _metadata.Purpose : purpose;
        var requestedScopes = MergeScopes(scopes);

        var cacheKey = AuthzCache.Key(PrincipalKey(), RenderResource(resource), action, effectivePurpose, requestedScopes);
        var cached = _cache.TryGet(cacheKey);
        if (cached is not null)
        {
            return (cached.Allowed, cached);
        }

        var request = BuildAuthzRequest(resource, action, effectivePurpose, requestedScopes);
        var decision = await AuthorizeAsync(request, ct);
        _cache.Set(cacheKey, decision);
        return (decision.Allowed, decision);
    }

    /// <summary>
    /// Like <see cref="CanAsync"/> but throws <see cref="UdbAuthzDeniedException"/>
    /// on deny. Returns the allowing decision on success.
    /// </summary>
    public async Task<AuthzV1.Decision> RequireAsync(
        AuthzV1.ResourceRef resource, string action, string purpose = "",
        IEnumerable<string>? scopes = null, CancellationToken ct = default)
    {
        var (allowed, decision) = await CanAsync(resource, action, purpose, scopes, ct);
        if (!allowed)
        {
            throw new UdbAuthzDeniedException(RenderResource(resource), action, decision);
        }
        return decision;
    }

    /// <summary>
    /// Returns the raw decision for a check without throwing — useful for UIs
    /// that surface the deny reason. Does not consult or populate the cache so
    /// the explanation is always fresh.
    /// </summary>
    public async Task<AuthzV1.Decision> ExplainAsync(
        AuthzV1.ResourceRef resource, string action, string purpose = "",
        IEnumerable<string>? scopes = null, CancellationToken ct = default)
    {
        var effectivePurpose = string.IsNullOrEmpty(purpose) ? _metadata.Purpose : purpose;
        var request = BuildAuthzRequest(resource, action, effectivePurpose, MergeScopes(scopes));
        return await AuthorizeAsync(request, ct);
    }

    /// <summary>
    /// Batch-check many (object, action) pairs for the configured user in one
    /// RPC (BatchCheckPermissions). Returns a map keyed by <c>"object:action"</c>
    /// to the boolean allow result, mirroring the server response map keys.
    /// </summary>
    public async Task<IReadOnlyDictionary<string, bool>> BatchCanAsync(
        IEnumerable<(string Object, string Action)> checks, string domain = "", CancellationToken ct = default)
    {
        var request = new AuthzV1.BatchCheckPermissionsRequest
        {
            UserId = _metadata.UserId,
            Domain = string.IsNullOrEmpty(domain) ? _metadata.TenantId : domain,
        };
        foreach (var (obj, action) in checks)
        {
            request.Checks.Add(new AuthzV1.PermissionCheck { Object = obj, Action = action });
        }
        var resp = await _authz.BatchCheckPermissionsAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
        return new Dictionary<string, bool>(resp.Results);
    }

    // ── Stage 2: native database fast-path access (item 138) ──────────────────
    /// <summary>
    /// Authorize and, when allowed, return the native-access grant (restricted
    /// role + scoped DSN + RLS session variables). Returns <c>null</c> when
    /// access is allowed but no grant was minted; throws on deny.
    /// <paramref name="scopes"/> augments the metadata scopes for the request's
    /// <c>requested_scopes</c>.
    /// </summary>
    public async Task<AuthzV1.NativeAccessGrant?> NativeAccessAsync(
        AuthzV1.ResourceRef resource, string action, string purpose = "",
        IEnumerable<string>? scopes = null, CancellationToken ct = default)
    {
        var principal = BuildPrincipal();
        var request = new AuthzV1.NativeAccessRequest
        {
            Principal = principal,
            TenantId = _metadata.TenantId,
            ProjectId = _metadata.ProjectId,
            Resource = resource,
            Action = action,
            Purpose = string.IsNullOrEmpty(purpose) ? _metadata.Purpose : purpose,
        };
        request.RequestedScopes.AddRange(MergeScopes(scopes));
        var resp = await _authz.GetNativeAccessAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
        if (resp.Decision is { Allowed: false })
        {
            throw new UdbAuthzDeniedException(RenderResource(resource), action, resp.Decision);
        }
        return resp.Grant;
    }

    // ── Stage 2: signed policy bundle (item 140) ──────────────────────────────
    public async Task<AuthzV1.SignedPolicyBundle> GetPolicyBundleAsync(CancellationToken ct = default)
    {
        var request = new AuthzV1.PolicyBundleRequest
        {
            TenantId = _metadata.TenantId,
            ProjectId = _metadata.ProjectId,
        };
        var resp = await _authz.GetPolicyBundleAsync(request, Headers(), cancellationToken: ct).ResponseAsync;
        if (!string.IsNullOrEmpty(_bundleSecret))
        {
            VerifyPolicyBundle(resp.Bundle, _bundleSecret!);
        }
        return resp.Bundle;
    }

    /// <summary>
    /// Verify the HMAC-SHA256 signature of a <see cref="AuthzV1.SignedPolicyBundle"/>.
    /// The server signs the <c>bundle</c> bytes and emits the signature as
    /// lowercase hex (the proto comment says Base64, but the server emits hex —
    /// we match the server). Comparison is constant-time. Throws
    /// <see cref="UdbPolicyBundleSignatureException"/> on mismatch.
    /// </summary>
    public static void VerifyPolicyBundle(AuthzV1.SignedPolicyBundle signed, string secret)
    {
        ArgumentNullException.ThrowIfNull(signed);
        if (string.IsNullOrEmpty(secret))
        {
            throw new ArgumentException("bundle secret must be non-empty", nameof(secret));
        }

        var expected = HmacSha256HexLower(secret, signed.Bundle.ToByteArray());
        var actual = signed.Signature ?? string.Empty;

        // Fixed-time compare over the raw bytes of the lowercase-hex strings.
        var expectedBytes = Encoding.ASCII.GetBytes(expected);
        var actualBytes = Encoding.ASCII.GetBytes(actual);
        if (!CryptographicOperations.FixedTimeEquals(expectedBytes, actualBytes))
        {
            throw new UdbPolicyBundleSignatureException(signed.KeyId, signed.Algorithm);
        }
    }

    /// <summary>Lowercase-hex HMAC-SHA256 of <paramref name="message"/> under <paramref name="secret"/>.</summary>
    private static string HmacSha256HexLower(string secret, byte[] message)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var mac = hmac.ComputeHash(message);
        return Convert.ToHexString(mac).ToLowerInvariant();
    }

    // ── helpers ───────────────────────────────────────────────────────────────
    private AuthzV1.Principal BuildPrincipal()
    {
        var principal = new AuthzV1.Principal
        {
            UserId = _metadata.UserId,
            ServiceIdentity = _metadata.ServiceIdentity,
            TenantId = _metadata.TenantId,
            ProjectId = _metadata.ProjectId,
        };
        principal.Scopes.AddRange(_metadata.Scopes);
        return principal;
    }

    private AuthzV1.AuthzRequest BuildAuthzRequest(
        AuthzV1.ResourceRef resource, string action, string purpose, IReadOnlyList<string> requestedScopes)
    {
        var request = new AuthzV1.AuthzRequest
        {
            Principal = BuildPrincipal(),
            TenantId = _metadata.TenantId,
            ProjectId = _metadata.ProjectId,
            Resource = resource,
            Action = action,
            Purpose = purpose,
        };
        request.RequestedScopes.AddRange(requestedScopes);
        return request;
    }

    /// <summary>Union of the metadata scopes and any per-call scopes, de-duplicated.</summary>
    private IReadOnlyList<string> MergeScopes(IEnumerable<string>? extra)
    {
        if (extra is null)
        {
            return _metadata.Scopes;
        }
        var merged = new List<string>(_metadata.Scopes);
        foreach (var s in extra)
        {
            if (!string.IsNullOrEmpty(s) && !merged.Contains(s))
            {
                merged.Add(s);
            }
        }
        return merged;
    }

    private string PrincipalKey()
        => $"{_metadata.TenantId}/{_metadata.ProjectId}/{_metadata.UserId}/{_metadata.ServiceIdentity}";

    /// <summary>Stable, human-readable rendering of a resource for cache keys / errors.</summary>
    private static string RenderResource(AuthzV1.ResourceRef r)
    {
        if (!string.IsNullOrEmpty(r.MessageType)) return r.MessageType;
        if (!string.IsNullOrEmpty(r.Table)) return string.IsNullOrEmpty(r.Schema) ? r.Table : $"{r.Schema}.{r.Table}";
        if (!string.IsNullOrEmpty(r.Collection)) return r.Collection;
        if (!string.IsNullOrEmpty(r.ResourceName)) return r.ResourceName;
        if (!string.IsNullOrEmpty(r.Path)) return r.Path;
        if (!string.IsNullOrEmpty(r.ResourceType)) return r.ResourceType;
        return r.ResourceId;
    }

    public ValueTask DisposeAsync()
    {
        _ownedChannel?.Dispose();
        return ValueTask.CompletedTask;
    }
}

File: sdk/java/src/main/java/dev/udb/client/UdbAuthClient.java
package dev.udb.client;

import com.udb.core.authn.services.v1.AuthnRequest;
import com.udb.core.authn.services.v1.AuthnResponse;
import com.udb.core.authn.services.v1.AuthnServiceGrpc;
import com.udb.core.authz.services.v1.AuthzRequest;
import com.udb.core.authz.services.v1.AuthzServiceGrpc;
import com.udb.core.authz.services.v1.BatchCheckPermissionsRequest;
import com.udb.core.authz.services.v1.BatchCheckPermissionsResponse;
import com.udb.core.authz.services.v1.CheckAccessRequest;
import com.udb.core.authz.services.v1.CheckAccessResponse;
import com.udb.core.authz.services.v1.Decision;
import com.udb.core.authz.services.v1.NativeAccessGrant;
import com.udb.core.authz.services.v1.NativeAccessRequest;
import com.udb.core.authz.services.v1.NativeAccessResponse;
import com.udb.core.authz.services.v1.PermissionCheck;
import com.udb.core.authz.services.v1.PolicyBundleRequest;
import com.udb.core.authz.services.v1.Principal;
import com.udb.core.authz.services.v1.ResourceRef;
import com.udb.core.authz.services.v1.SignedPolicyBundle;
import io.grpc.ManagedChannel;
import io.grpc.Metadata;
import io.grpc.stub.MetadataUtils;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.List;
import java.util.Map;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Hand-written auth ergonomics over the generated AuthnService / AuthzService
 * stubs, mirroring {@link UdbClient}'s metadata convention (item 111).
 */
public class UdbAuthClient implements AutoCloseable {
  private final ManagedChannel channel;
  private final UdbMetadata metadata;
  private final UdbCredentials credentials;
  private final AuthnServiceGrpc.AuthnServiceBlockingStub authn;
  private final AuthzServiceGrpc.AuthzServiceBlockingStub authz;

  /**
   * Optional shared secret used to verify the HMAC-SHA256 signature on policy
   * bundles fetched via {@link #getPolicyBundle()}. When non-null/non-empty, an
   * unverifiable bundle raises {@link UdbPolicyBundleSignatureException}.
   */
  private volatile String bundleSecret;

  public UdbAuthClient(String target, UdbMetadata metadata) {
    this(UdbChannels.forTarget(target, false), metadata);
  }

  public UdbAuthClient(ManagedChannel channel, UdbMetadata metadata) {
    this(channel, metadata, UdbCredentials.fromMetadata(metadata));
  }

  /** Build over a shared, mutable credentials holder so a refreshed token is sent
   *  on the next authn/authz call without rebuilding the channel. */
  public UdbAuthClient(ManagedChannel channel, UdbMetadata metadata, UdbCredentials credentials) {
    this.channel = channel;
    this.metadata = metadata;
    this.credentials = credentials;
    this.authn = AuthnServiceGrpc.newBlockingStub(channel);
    this.authz = AuthzServiceGrpc.newBlockingStub(channel);
  }

  /**
   * Channel-free constructor for subclassed test doubles that override the RPC
   * surface ({@link #can}, {@link #authorize}, etc.) and never touch the wire.
   * No channel/stubs are created, so the network RPC methods must not be called
   * on instances built this way.
   */
  protected UdbAuthClient(UdbMetadata metadata) {
    this.channel = null;
    this.metadata = metadata;
    this.credentials = UdbCredentials.fromMetadata(metadata);
    this.authn = null;
    this.authz = null;
  }

  private Metadata headers() {
    return UdbClient.headers(metadata, credentials.bearerToken(), credentials.apiKey());
  }

  /** The shared credentials holder backing this client's outbound auth headers. */
  public UdbCredentials credentials() {
    return credentials;
  }

  // ── Authentication ────────────────────────────────────────────────────────
  public AuthnResponse authenticate(AuthnRequest request) {
    return authn.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers())).authenticate(request);
  }

  public AuthnResponse authenticateBearer(String token) {
    return authenticate(AuthnRequest.newBuilder()
        .setBearerToken(token)
        .setTenantHint(metadata.tenantId())
        .setProjectHint(metadata.projectId())
        .build());
  }

  public AuthnResponse authenticateApiKey(String apiKey) {
    return authenticate(AuthnRequest.newBuilder()
        .setApiKey(apiKey)
        .setTenantHint(metadata.tenantId())
        .setProjectHint(metadata.projectId())
        .build());
  }

  public AuthnResponse authenticateSession(String sessionId) {
    return authenticate(AuthnRequest.newBuilder()
        .setSessionId(sessionId)
        .setTenantHint(metadata.tenantId())
        .setProjectHint(metadata.projectId())
        .build());
  }

  // ── Authorization ───────────────────────────────────────────────────────────
  public Decision authorize(AuthzRequest request) {
    return authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()))
        .authorize(request)
        .getDecision();
  }

  /**
   * Authorize the bound principal acting on a resource. The caller's metadata
   * scopes are forwarded as {@code requested_scopes} so the authority can apply
   * scope-narrowing exactly as the Go/Python/TS SDKs do.
   */
  public Decision can(ResourceRef resource, String action, String purpose) {
    Principal principal = boundPrincipal();
    AuthzRequest request = AuthzRequest.newBuilder()
        .setPrincipal(principal)
        .setTenantId(metadata.tenantId())
        .setProjectId(metadata.projectId())
        .setResource(resource)
        .setAction(action)
        .setPurpose(purpose == null || purpose.isEmpty() ? metadata.purpose() : purpose)
        .addAllRequestedScopes(metadata.scopes())
        .build();
    return authorize(request);
  }

  /**
   * Authorize and throw {@link UdbAuthzDeniedException} when the decision denies
   * access; returns the allow {@link Decision} otherwise (matched policies,
   * required scopes, server cache TTL).
   */
  public Decision require(ResourceRef resource, String action, String purpose) {
    Decision decision = can(resource, action, purpose);
    if (!decision.getAllowed()) {
      throw new UdbAuthzDeniedException(resource, action, decision);
    }
    return decision;
  }

  /**
   * Non-throwing companion to {@link #require}: returns the full {@link Decision}
   * (allowed flag + deny reason + matched policies) for inspection.
   */
  public Decision explain(ResourceRef resource, String action, String purpose) {
    return can(resource, action, purpose);
  }

  /**
   * Forward a Casbin-style (user, domain, object, action) request and report
   * whether access is allowed.
   */
  public CheckAccessResponse checkAccess(CheckAccessRequest request) {
    return authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()))
        .checkAccess(request);
  }

  /**
   * Evaluate many (object, action) checks in one round-trip via the
   * BatchCheckPermissions RPC, returning the server's per-check
   * {@code object:action -> allowed} result map.
   */
  public Map<String, Boolean> batchCan(List<PermissionCheck> checks) {
    BatchCheckPermissionsRequest request = BatchCheckPermissionsRequest.newBuilder()
        .setUserId(metadata.userId())
        .setDomain(metadata.tenantId())
        .addAllChecks(checks)
        .build();
    BatchCheckPermissionsResponse resp =
        authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()))
            .batchCheckPermissions(request);
    return resp.getResultsMap();
  }

  /** Build a single {@link PermissionCheck} for use with {@link #batchCan}. */
  public static PermissionCheck check(String object, String action) {
    return PermissionCheck.newBuilder().setObject(object).setAction(action).build();
  }

  private Principal boundPrincipal() {
    return Principal.newBuilder()
        .setUserId(metadata.userId())
        .setServiceIdentity(metadata.serviceIdentity())
        .setTenantId(metadata.tenantId())
        .setProjectId(metadata.projectId())
        .addAllScopes(metadata.scopes())
        .build();
  }

  /** The metadata identity bound to this client (used by {@link AuthzCache}). */
  public UdbMetadata metadata() {
    return metadata;
  }

  /** Raw AuthnService stub with caller headers attached. */
  public AuthnServiceGrpc.AuthnServiceBlockingStub authn() {
    return authn.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()));
  }

  /** Raw AuthzService stub with caller headers attached. */
  public AuthzServiceGrpc.AuthzServiceBlockingStub authz() {
    return authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()));
  }

  // ── Stage 2: native database fast-path access (item 138) ────────────────────
  /**
   * Authorize and, when allowed, return the native-access grant (restricted role
   * + scoped DSN + RLS session variables). Returns {@code null} when access is
   * allowed but no grant was minted; throws when the decision denies access.
   */
  public NativeAccessGrant nativeAccess(ResourceRef resource, String action, String purpose) {
    Principal principal = boundPrincipal();
    NativeAccessRequest request = NativeAccessRequest.newBuilder()
        .setPrincipal(principal)
        .setTenantId(metadata.tenantId())
        .setProjectId(metadata.projectId())
        .setResource(resource)
        .setAction(action)
        .setPurpose(purpose == null || purpose.isEmpty() ? metadata.purpose() : purpose)
        .addAllRequestedScopes(metadata.scopes())
        .build();
    NativeAccessResponse resp = authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()))
        .getNativeAccess(request);
    if (resp.hasDecision() && !resp.getDecision().getAllowed()) {
      throw new IllegalStateException("udb: native access denied: " + resp.getDecision().getDenyReason());
    }
    return resp.hasGrant() ? resp.getGrant() : null;
  }

  // ── Stage 2: signed policy bundle (item 140) ────────────────────────────────
  /**
   * Set (or clear with {@code null}/empty) the shared secret used to verify the
   * HMAC-SHA256 signature on bundles fetched by {@link #getPolicyBundle()}. With a
   * secret configured, {@link #getPolicyBundle()} throws {@link
   * UdbPolicyBundleSignatureException} on a signature mismatch. Returns {@code
   * this} for chaining.
   */
  public UdbAuthClient withBundleSecret(String secret) {
    this.bundleSecret = secret;
    return this;
  }

  /**
   * Fetch the signed policy bundle for the bound tenant/project. When a bundle
   * secret has been configured via {@link #withBundleSecret(String)}, the bundle's
   * HMAC-SHA256 signature is verified before it is returned; a mismatch raises
   * {@link UdbPolicyBundleSignatureException}.
   */
  public SignedPolicyBundle getPolicyBundle() {
    PolicyBundleRequest request = PolicyBundleRequest.newBuilder()
        .setTenantId(metadata.tenantId())
        .setProjectId(metadata.projectId())
        .build();
    SignedPolicyBundle bundle =
        authz.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers()))
            .getPolicyBundle(request)
            .getBundle();
    String secret = this.bundleSecret;
    if (secret != null && !secret.isEmpty() && !verifyPolicyBundle(bundle, secret)) {
      throw new UdbPolicyBundleSignatureException(
          "udb: policy bundle signature verification failed"
              + (bundle.getKeyId().isEmpty() ? "" : " (key_id=" + bundle.getKeyId() + ")"),
          bundle);
    }
    return bundle;
  }

  /**
   * Recompute the HMAC-SHA256 over {@code signed.getBundle()} bytes using {@code
   * secret} and constant-time compare the lowercase-hex digest to {@code
   * signed.getSignature()}. The server emits a lowercase-hex HMAC (the proto
   * comment's "Base64" is stale), so this matches the wire format. Returns {@code
   * true} when the signature is valid.
   */
  public static boolean verifyPolicyBundle(SignedPolicyBundle signed, String secret) {
    if (signed == null || secret == null || secret.isEmpty()) {
      return false;
    }
    String expected = hmacSha256Hex(signed.getBundle().toByteArray(), secret);
    return constantTimeEquals(expected, signed.getSignature());
  }

  private static String hmacSha256Hex(byte[] data, String secret) {
    try {
      Mac mac = Mac.getInstance("HmacSHA256");
      mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
      byte[] digest = mac.doFinal(data);
      StringBuilder sb = new StringBuilder(digest.length * 2);
      for (byte b : digest) {
        sb.append(Character.forDigit((b >> 4) & 0xF, 16));
        sb.append(Character.forDigit(b & 0xF, 16));
      }
      return sb.toString();
    } catch (java.security.GeneralSecurityException err) {
      throw new IllegalStateException("udb: HMAC-SHA256 unavailable", err);
    }
  }

  /** Length-independent constant-time string compare (via {@link MessageDigest}). */
  private static boolean constantTimeEquals(String a, String b) {
    byte[] x = a == null ? new byte[0] : a.getBytes(StandardCharsets.UTF_8);
    byte[] y = b == null ? new byte[0] : b.getBytes(StandardCharsets.UTF_8);
    return MessageDigest.isEqual(x, y);
  }

  @Override
  public void close() {
    if (channel != null) {
      channel.shutdownNow();
    }
  }
}
